File: app/src/main/java/piotr/michalkiewicz/mealplannerclient/user/service_generator/UserServiceGenerator.java
package piotr.michalkiewicz.mealplannerclient.user.service_generator;

import android.content.Context;

import piotr.michalkiewicz.mealplannerclient.auth.ServiceGenerator;
import piotr.michalkiewicz.mealplannerclient.user.model.UserAccount;
import piotr.michalkiewicz.mealplannerclient.user.model.UserSettings;
import retrofit2.Call;
import retrofit2.Callback;

import static piotr.michalkiewicz.mealplannerclient.utils.ConstantValues.PASSWORD_GRANT_TYPE;

public class UserServiceGenerator {

    private ServiceGenerator apiClient;
    private UserService userService;

    public UserServiceGenerator(){
        apiClient = new ServiceGenerator();
    }

    public void getUserAccount(Callback<UserAccount> callback){
        initUserServiceIfNull();
        Call<UserAccount> callAsync = userService.getUserAccount();

        callAsync.enqueue(callback);
    }

    public void signUp(UserAccount userAccount, Callback<UserAccount> callback){
        SignUpService signUpService = apiClient.getSignUpApi();
        Call<UserAccount> callAsync = signUpService.signUp(userAccount.getUsername(),
                userAccount.getEmail(), userAccount.getPassword(), PASSWORD_GRANT_TYPE);

        callAsync.enqueue(callback);
    }

    public void saveUserAccountData(UserAccount data, Callback<UserAccount> callback){
        initUserServiceIfNull();
        Call<UserAccount> callAsync = userService.updateAccount(data);

        callAsync.enqueue(callback);
    }

    public void saveUserSettings(UserSettings userSettings, Callback<UserAccount> callback){
        initUserServiceIfNull();
        Call<UserAccount> callAsync = userService.updateSettings(userSettings);

        callAsync.enqueue(callback);
    }

    private void initUserServiceIfNull(){
        if(userService==null) {
            userService = apiClient.getUserApi();
        }
    }
}

File: app/src/main/java/piotr/michalkiewicz/mealplannerclient/user/service_generator/UserServiceGenerator.kt
package piotr.michalkiewicz.mealplannerclient.user.service_generator

import android.util.Log
import io.reactivex.schedulers.Schedulers
import piotr.michalkiewicz.mealplannerclient.auth.ServiceGenerator
import piotr.michalkiewicz.mealplannerclient.user.model.UserAccount
import piotr.michalkiewicz.mealplannerclient.user.model.UserPreference
import piotr.michalkiewicz.mealplannerclient.user.model.UserSettings
import retrofit2.Callback

class UserServiceGenerator : ServiceGenerator() {

    private lateinit var userService: UserService

    init {
        if (!::userService.isInitialized) {
            val retrofit = retrofitBuilder()
            userService = retrofit.create(UserService::class.java)
        }
    }

    fun getUserAccount(callback: Callback<UserAccount>) {  //toDo delete ?
        val callAsync = userService.getUserAccount()
        callAsync.enqueue(callback)
    }

    fun signUp(userAccount: UserAccount, callback: Callback<UserAccount>) {
        val callAsync = userService.signUp(userAccount)
        callAsync.enqueue(callback)
    }

    fun saveUserAccountData(data: UserAccount, callback: Callback<UserAccount>) {  //toDo delete ?
        val callAsync = userService.editAccountSettings(data)
        callAsync.enqueue(callback)
    }

    fun updateUserSettings(settings: UserSettings) {                                                //toDO delete ?
        val updateUserSettingsAPI = userService.updateUserSettings(settings)
        updateUserSettingsAPI.subscribeOn(Schedulers.io())
                .subscribe({ result ->
                    Log.i("updateUserSettings", result.toString())
                }, { error ->
                    Log.i("error", error.toString())
                })
    }


    fun updateUserPreference(userPreference: UserPreference){
        val userCustomizationEndPoint = userService.updateUserPreference(userPreference)
        userCustomizationEndPoint.subscribeOn(Schedulers.io())
                .subscribe({result ->
                    Log.i("userPreferenceCustomiz", result.toString())
                }, { error ->
                    Log.i("error", error.toString())
                })
    }
}